### PR TITLE
Bug Fixes for Issue #1088

### LINKED ISSUE #1088: TableFilter Programmatic API Bugs
**[Original report](https://bitbucket.org/controlsfx/controlsfx/issue/897) by Thomas Nield (Bitbucket: [thomasnield](https://bitbucket.org/thomasnield), GitHub: [thomasnield](https://github.com/thomasnield)).**

----------------------------------------

A number of fixes need to be done to make the programmatic filtering API work properly on the TableFilter

1) selectValue(), unselectValue(), selectAllValues(), unselectAllValues() need to initialize their respective `ColumnFilter` the moment they are called

2) The icon application/column width bump procedure needs to be moved from the `ColumnFilter#applyFilter()` to `TableFilter#executeFilter()` so the icon shows when a filter is applied programmatically

### PR DESCRIPTION
**Pull request by Mason-Chou**
These changes should address the bugs mentioned in #1088 

Fix #1088

### CODE REVIEW

File: controlsfx/src/main/java/org/controlsfx/control/table/ColumnFilter.java [hunk 1]
```diff
@@ -135,6 +147,23 @@ FilterPanel getFilterPanel() {
...
+     */
+    void applyFilterIcon() {
+        if (hasUnselections()) {
+            tableColumn.setGraphic(filterImageView.get());
+            if (!bumpedWidth) {
+                tableColumn.setPrefWidth(tableColumn.getWidth() + 20);
+                bumpedWidth = true;
+            }
+        }
+        else {
```
Comment (Maxoudela) [inline controlsfx/src/main/java/org/controlsfx/control/table/ColumnFilter.java]: I don't think we follow that pattern. The `else statement` should be on the same line.

File: controlsfx/src/main/java/org/controlsfx/control/table/FilterPanel.java [hunk 2]
```diff
@@ -62,22 +59,12 @@
     private final FilteredList<FilterValue> filterList;
     private final TextField searchBox = new TextField();
     private boolean searchMode = false;
```
Comment (Maxoudela) [inline controlsfx/src/main/java/org/controlsfx/control/table/FilterPanel.java]: This is only used in `initializeListeners`, maybe remove it?

Comment (Mason-Chou) [reply controlsfx/src/main/java/org/controlsfx/control/table/FilterPanel.java]: Removed ```searchMode``` entirely as it wasn't being used even in ```initializedListeners```.

**Comment by abhinayagarwal** [comment]
Hi @Mason-Chou,

As @thomasnield is busy, I will try to review this PR, if you are still willing to work on this. Can you fix the conflicts?

**Comment by Mason-Chou** [comment]
@ abhinayagarwal FYI, fixed the merge conflicts.

**Comment by danielpeintner** [comment]
With this PR (or yet another one) we might also want to extend the [HelloTableFilter](https://github.com/controlsfx/controlsfx/blob/master/controlsfx-samples/src/main/java/org/controlsfx/samples/tableview/HelloTableFilter.java) sample to show how programatically setting filters on a TableView work...

**Comment by danielpeintner** [comment]
Thanks @Mason-Chou for your work.

I do not have review or approve rights but I did look into your PR and it seems working fine in general but I did run into some issues.

I updated [HelloTableFilter](https://github.com/controlsfx/controlsfx/blob/master/controlsfx-samples/src/main/java/org/controlsfx/samples/tableview/HelloTableFilter.java) sample to show and explore some possible scenarios.

What I found not working
* Filtering of LastName column programmatically works while the other 2 columns do not work properly (I do not really understand why?)
* Filtering by clicking on table column header (e.g., Firstname "Isabella") and than resetting filter by clicking the button "Reset All" keeps filter icon, filtered values are reset though
* Sometimes fields are greyed out but checkbox is still selectable (e.g. after the action/bullet list item before)

Do you want to proceed with the PR? I am willing to help.
I attached the example I used.

[HelloTableFilter.zip](https://github.com/controlsfx/controlsfx/files/4298782/HelloTableFilter.zip)

**Comment by Mason-Chou** [comment]
@danielpeintner I haven't worked on this for a while, nut I know the answer to one of your questions off the top of my head. The other two I can look into over the weekend. The PR should continue as it does address the two problem #1088 has. 

> Filtering of LastName column programmatically works while the other 2 columns do not work properly (I do not really understand why?)

`selectValue()` does an exact match. It does not try to match values that contains substring. This behavior isn't something that I've introduced through this PR, but is currently present in the original codebase. I'm not sure the reasoning behind this, but I believe it's tied to the ColumnFilters using Generics and thus needing to use the `Object.equals()` method for equality. It might be wise to change this behavior, but that's out of the scope of this issue.

**Comment by danielpeintner** [comment]
> selectValue() does an exact match. It does not try to match values which could contains the given substring. 

Thanks for your prompt reply. You are right it needs to be an exact match.

I might have mixed it up because the filter in the table column does partial matching using the *default* search strategy .

I am not sure whether changing the behavior is the way to go or introducing separate method that applies the set (`setSearchStrategy(...)`) makes more sense...

**Comment by Maxoudela** [comment]
> 
> 
> Thanks @Mason-Chou for your work.
> 
> I do not have review or approve rights but I did look into your PR and it seems working fine in general but I did run into some issues.
> 
> I updated [HelloTableFilter](https://github.com/controlsfx/controlsfx/blob/master/controlsfx-samples/src/main/java/org/controlsfx/samples/tableview/HelloTableFilter.java) sample to show and explore some possible scenarios.
> 
> What I found not working
> 
>     * Filtering of LastName column programmatically works while the other 2 columns do not work properly (I do not really understand why?)
> 
>     * Filtering by clicking on table column header (e.g., Firstname "Isabella") and than resetting filter by clicking the button "Reset All" keeps filter icon, filtered values are reset though
> 
>     * Sometimes fields are greyed out but checkbox is still selectable (e.g. after the action/bullet list item before)
> 
> 
> Do you want to proceed with the PR? I am willing to help.
> I attached the example I used.
> 
> [HelloTableFilter.zip](https://github.com/controlsfx/controlsfx/files/4298782/HelloTableFilter.zip)

Are these regression introduced by this PR, or pre-existing issues?

**Comment by danielpeintner** [comment]
I did try the same steps with the old code and the same issues exist already today.

However, the current code is not really meant for programmatically changing the filter.. e.g.,

* the filter icon did never show up (this works with the PR now)
* the greyed out issue existed before (and still exists)

**Comment by danielpeintner** [comment]
I looked into the issue I previously highlighted and fired a PR to the existing repo from @Mason-Chou (see https://github.com/Mason-Chou/controlsfx/pull/2).

The PR contains
* updated hello example to show how one can programmatically use the filter
* fixed the *reset all* issues by introducing a new method `resetAllFilters()`
(touching the existing `resetFilter()` didn't seem right to me given it is used by `ColumnFilter` and I think fulfills a different purpose, it just sets the predicates to true)

**Comment by thomasnield** [comment]
Hey all, I'm sorry I've been off the grid so much. The nature of my work has been changing a lot this year. If I have time I'll get to this. But please don't let me hold up any progress in improving the `TableFilter`.

**Comment by danielpeintner** [comment]
@thomasnield thank you for your feedback.

I did some more testing and found one remaining issue with search.

* enter "J" (in my updated HelloTabelFilter sample) in name search column and press "apply" selects the right columns. However returning to the filter "J" is still shown in the search field and removing the character shows as if all names are selected

I am willing to provide a fix/work on it but let's wait till @Mason-Chou has time to look at my other changes....

I think we should do further testing...

**Comment by Mason-Chou** [comment]
@danielpeintner just merged you changes in, i'll work on getting the merge conflicts fixed soon

**Comment by abhinayagarwal** [comment]
Please target the PR to `jfx-13` branch.

**Comment by Mason-Chou** [comment]
@abhinayagarwal base branch changed with merge conflicts fixed. The Travis CI build is failing due to a permission issue, can confirm the build runs on my machine with Java 11

**Comment by abhinayagarwal** [comment]
Hi @Mason-Chou !

Pushed a fix to the base branch, please update PR.

**Comment by danielpeintner** [comment]
@abhinayagarwal @Maxoudela  is there anything that prevents this PR being merged?
IF *other* issues pop-up we can still resolve them step by step, right?

File: controlsfx-samples/src/main/java/org/controlsfx/samples/tableview/HelloTableFilter.java [hunk 3]
```diff
@@ -29,7 +29,10 @@
 import javafx.beans.property.SimpleStringProperty;
 import javafx.scene.Node;
 import javafx.scene.control.*;
+import javafx.scene.control.Button;
+import javafx.scene.control.Label;
 import javafx.scene.control.cell.PropertyValueFactory;
+import javafx.scene.layout.*;
```
Comment (Maxoudela) [inline controlsfx-samples/src/main/java/org/controlsfx/samples/tableview/HelloTableFilter.java:35]: I think that we avoid the "*" imports

**Review by Maxoudela — APPROVED** [review]

**Comment by Maxoudela** [comment]
I look quickly. Seems good apart from an "*" import that, I think, is not a good practice. But since it's in the sample, maybe it's not that important.

**Comment by abhinayagarwal** [comment]
Please remove the changes made to `gradlew` and `.project` files. Also, please update the LICENSE header to 2020 in respective files.

**Comment by Mason-Chou** [comment]
@abhinayagarwal see the new updates, should match with your standards.

File: controlsfx/src/main/java/org/controlsfx/control/table/FilterPanel.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2015, 2019, ControlsFX
+ * Copyright (c) 2015, 2016, 2020 ControlsFX
```
Comment (abhinayagarwal) [inline controlsfx/src/main/java/org/controlsfx/control/table/FilterPanel.java:2]: This should be updated to "Copyright (c) 2015, 2020, ControlsFX"

Comment (Mason-Chou) [reply controlsfx/src/main/java/org/controlsfx/control/table/FilterPanel.java]: fixed now.

**Review by abhinayagarwal — APPROVED** [review]

**Comment by abhinayagarwal** [comment]
This PR looks good. One last change: Please remove the file `.ci/deploy-snapshot.sh` from the PR. I will merge the PR once this has been updated.

**Pull request merged**